src/main/java/org/codehaus/groovy/reflection/ReflectionUtils.java

```diff
@@ -190,6 +191,15 @@ public static boolean trySetAccessible(final AccessibleObject ao) {
         return false;
     }
 
+    public static boolean hasAnnotationWithSimpleName(AccessibleObject obj, String simpleName) {
```

Comment:
Both `@groovy.annotations.Pure` and `@foo.bar.Pure` have same simple name `Pure`. If user uses `@foo.bar.Pure` to mark methods, which will be treated as "pure" methods too.

So I recommend using full class name instead of simple name.

Reply:
We can use full names. There are other annotations, e.g.:
checker framework's `org.checkerframework.dataflow.qual.Pure`
pure4j's `org.pure4j.annotations.pure.Pure`
JPure's `@Pure`
JML's `@Pure`
XText has `org.eclipse.xtext.xbase.lib.Pure`
Effect tracked Java uses `@Pure` (and `@Impure`)
C4J uses `de.vksi.c4j.Pure`
and `net.digitalid.utility.annotations.method.Pure`
which do essentially the same thing as ours.
Counter examples: Frege has a `pure` keyword, JetBrains has `@Contract` with a `pure` boolean annotation attribute.

Reply:
I haven't worked out a way to make it extensible without making it ugly.

And my next goal will be to write a purity type checker extension which would hopefully be able to be made to work with any of the above. Although I can think of a few options for this second use case.